luxtronik/__init__.py

```diff
@@ -24,28 +47,44 @@ def __init__(self, host, port, safe=True):
         self.visibilities = Visibilities()
         self.read()
 
-    def _connect(self):
-        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        self._socket.connect((self._host, self._port))
-        LOGGER.info("Connected to Luxtronik heatpump %s:%s", self._host, self._port)
-
-    def _disconnect(self):
-        self._socket.close()
-        LOGGER.info(
-            "Disconnected from Luxtronik heatpump  %s:%s", self._host, self._port
-        )
+    def __del__(self):
+        if self._socket is not None:
+            if not is_socket_closed(self._socket):
+                self._socket.close()
+            self._socket = None
+            LOGGER.info(
+                "Disconnected from Luxtronik heatpump %s:%s", self._host, self._port
+            )
 
     def read(self):
         """Read data from heatpump."""
-        self._connect()
+        self._read_write(write=False)
+
+    def write(self):
+        """Write patameter to heatpump."""
+        self._read_write(write=True)
+
+    def _read_write(self, write=False):
+        # Ensure only one socket operation at the same time
+        with self._lock:
+            is_none = self._socket is None
+            if is_none:
+                self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            if is_none or is_socket_closed(self._socket):
+                self._socket.connect((self._host, self._port))
+                LOGGER.info(
+                    "Connected to Luxtronik heatpump %s:%s", self._host, self._port
+                )
+            if write:
+                return self._write()
+            self._read()
+
+    def _read(self):
         self._read_parameters()
         self._read_calculations()
         self._read_visibilities()
-        self._disconnect()
 
-    def write(self):
-        """Write patameter to heatpump."""
```

Comment:
Seems like you're removing the docstring here. Is this by accident or on purpose?

Reply:
accident - function changed from write to _write.
Write see L63 has docstring